### ISSUE TITLE
Issue recording Unicode characters from a host Windows C++ application

### ISSUE THREAD

**Issue by DianaI13** [MPSystemsServices/CodeBase-for-DBF#13]
Hi,

Do you know if Unicode characters are supported in database fields (dBase 4). We are trying to update the database from Windows C++ application.

Thank you in advance,
Diana

**Comment by addsuminc** [comment]
Yes.  See REF.PDF and the 6.5 reference:    https://github.com/MPSystemsServices/CodeBase-for-DBF/blob/master/readme.txt 

Tony Frates
Addsum Business Software, Inc.


----- Original Message -----
From: DianaI13 ***@***.***>
To: MPSystemsServices/CodeBase-for-DBF ***@***.***>
Cc: Subscribed ***@***.***>
Date: Fri, 29 Jul 2022 11:02:04 -0700
Subject: [MPSystemsServices/CodeBase-for-DBF] Issue recording Unicode characters from a host Windows C++ application (Issue #13)

> Hi,
> 
> Do you know if Unicode characters are supported in database fields (dBase 4). We are trying to update the database from Windows C++ application.
> 
> Thank you in advance,
> Diana
> 


-- 
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

**Comment by DianaI13** [comment]
Thank you, Tony for your prompt response.

I found the notes for Unicode.

My team will explore.

Best Regards,
Diana Ivanov
Software Development Manager @ Triconex/Schneider-Electric

**Comment by BobHocker** [comment]
Hello,

I'm working with Diana, above.

When we specify one or more "W" fields, the first byte of the generated DBF file changes from 0x03 (dBASE Level 5) to 0x30 which appears to be a FoxPro specific value. This means that we can no longer read the file with non-CBASE tools such as Excel. This corresponds to the CodeBase Reference Guide Appendix E text that says W fields "will not be recognized by other xBASE tools". So I have two questions.
1) Is there any way to alter the code to generate a type 0x03 dbf file with a "W" field that can be read by other tools?
2) If the above is not possible, then we are trying to use a FoxPro driver. Can you point me to an ODBC data source driver that will allow our application to read these files?

Thank you,
Bob Hocker

**Comment by BobHocker** [comment]
A follow-up question to my questions above:
We tried to open a dbf file that has a Unicode 'W' field in Microsoft Visual FoxPro, but it gave an error: "<the dbf file> is not a table". So even Visual FoxPro doesn't appear to recognize type 0x30 dbf files. Is this correct? Shouldn't Visual FoxPro recognize a type 0x30 dbf file written by Codebase? Does this mean that no product other than Codebase can read these files as dBase files? Thanks in advance for any information you can provide.

**Comment by addsuminc** [comment]
That seems to be what the documentation indicates on page 202 (of REF.PDF) because of that particular field type:

"Unicode fields store Unicode strings. The field is padded with nulls. When creating a table with this field type, remember to specify the length as twice the required number of Unicode characters, because each Unicode character is stored in 2 bytes.

This field type is only compatible with CodeBase. It will not be recognized by other xBASE tools. When creating a Unicode field in a FoxPro table, CODE4.compatibility must be set to 30."

We don't use FoxPro so can't otherwise be of much help.


Tony Frates
Addsum Business Software, Inc.


----- Original Message -----
From: BobHocker ***@***.***>
To: MPSystemsServices/CodeBase-for-DBF ***@***.***>
Cc: addsuminc ***@***.***>,  Comment ***@***.***>
Date: Wed, 24 Aug 2022 09:30:22 -0700
Subject: Re: [MPSystemsServices/CodeBase-for-DBF] Issue recording Unicode characters from a host Windows C++ application (Issue #13)

> A follow-up question to my questions above:
> We tried to open a dbf file that has a Unicode 'W' field in Microsoft Visual FoxPro, but it gave an error: "<the dbf file> is not a table". So even Visual FoxPro doesn't appear to recognize type 0x30 dbf files. Is this correct? Shouldn't Visual FoxPro recognize a type 0x30 dbf file written by Codebase? Does this mean that no product other than Codebase can read these files as dBase files? Thanks in advance for any information you can provide.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/MPSystemsServices/CodeBase-for-DBF/issues/13#issuecomment-1225960364
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>
> 

-- 
This email has been checked for viruses by Avast antivirus software.
www.avast.com

**Comment by BobHocker** [comment]
Thank you, Tony, for your response.
